An Example Function Code to find average of A list which contain values.

In [11]:
#A List Example
data1 = [70, 70, 70, 100, 100, 100, 120, 120, 150, 150,75,89,8,34,56,78,90,34,12,15,16,75,43,23,45,45,44,78,29,37,51,36,35,23,67,88,89,20,13,10,41,26,34,25,60,10,20,40,50,55,26,68,44,78]
data2 = [50, 60, 60, 50, 70, 70, 100, 80, 100, 90,43,23,45,94,51,36,35,23,67,88,89,20,13,10,41,26,34,25,60,10,20,40,50,55, 70, 100, 100, 100, 120, 12]

In [12]:
#How Much Value in every list ?

print("data1 consist of ", len(data1), "value")

print("data2 consist of ", len(data2), "value")

data1 consist of  54 value
data2 consist of  40 value


In [19]:
#Let Say we will to find how many values 100 in data1
data1.count(100)

3

In [30]:
#Finding Average in every list with Function
#1st Version
%%timeit
def average(data):
    jumlah = 0
    for item in data:
        jumlah += item
    average = jumlah/len(data)
    return average


48.1 ns ± 0.625 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [31]:
print('Average in data1:')
print(average(data1))
print('average in data2:')
print(average(data2))

Average in data1:
55.27777777777778
average in data2:
55.75


In [33]:
#2nd Version
%%timeit
def average(data):
    jumlah = 0
    for item in data:
        jumlah += 1
    average = sum(data)/jumlah
    return average


49.6 ns ± 1.11 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [34]:
print('Average in data1:')
print(average(data1))
print('average in data2:')
print(average(data2))

Average in data1:
55.27777777777778
average in data2:
55.75


Here i show you an interesting thing : 1st version faster than 2nd version

1st version : 48.1 ns ± 0.625 ns per loop


2nd version : 49.6 ns ± 1.11 ns per loop

In 1st version for loop fullfill the sum of the list first then average value counts last. Meanwhile, 2nd version, for loop counting how many values in list then average counts with sum of data divided by for loop results.

Interesting.

Hypotesis to learn :



1.   function with for loop elements more efficient if for loop is used to calculate instead of to count element.


2.   Better use len(data) to count element.




